### ISSUE TITLE
Question regarding pruning rate/ sparsity

### ISSUE THREAD

**Issue by ajithkumarmcw** [VainF/Torch-Pruning#214]
Thanks for a excellent repo, I performed two experiments on yolo model
exp1: prune rate = 0.5 , maxacc drop = 0.2, iteration steps = 16, map before pruning 0.68
Experiment got ended at 0.48 map 12th step with inference time of 0.89 seconds

exp2: prune rate = 0.5 , maxacc drop = 0.05, iteration steps = 16, map before pruning 0.68
Experiment got ended at 0.63 map 10th step with inference time of 1.1 seconds

Does it mean that sparsity of exp1's final model is more than exp2's final model or else both the experiments has same sparsity and only they are structurally pruned differently  . what would be the sparsity for exp1 and exp2
Is there a way to calculate sparsity after the model is pruned ?

**Comment by ajithkumarmcw** [comment]
@VainF could you kindly help me with estimating the sparsity of model after its pruned and finetuned. below code give 9.9e-5 even for a pruned and finetuned model.

```
def sparsity(model):
    # Return global model sparsity
    a, b = 0, 0
    for p in model.parameters():
        a += p.numel()
        b += (p == 0).sum()
    return b / a
```

**Comment by AymenBOUGUERRA** [comment]
Hello @ajithkumarmcw,

Can you please give more context regarding your issue ? such as script used / Yolo version ?
The author has included a print instruction in the scripts to mention the number of parameters and MACs (Multiplication Accumulation Cycle) before and after the pruning as shown here:
```
    base_macs, base_nparams = tp.utils.count_ops_and_params(model, example_inputs)
    pruner.step()
    pruned_macs, pruned_nparams = tp.utils.count_ops_and_params(model, example_inputs)
    print(model)
    print("Before Pruning: MACs=%f G, #Params=%f G"%(base_macs/1e9, base_nparams/1e9))
    print("After Pruning: MACs=%f G, #Params=%f G"%(pruned_macs/1e9, pruned_nparams/1e9))
```
leading to this output
![image](https://github.com/VainF/Torch-Pruning/assets/97101162/1977de62-5108-4006-a6b9-b232a0b933bb)

**Comment by AymenBOUGUERRA** [comment]
Hello again @ajithkumarmcw 

What this snippet of code does if I understood it correctly is to count the number of total parameters in the model then count the number of the parameters equal to 0 (sparsified) and then divide them to give you a sparsity ratio.

The result that you are having is completely normal and coherent in my opinion, and here is why:

- There is a big chaos in the naming of this approaches (sparse/prune) some people use sparse to refer to prune and vise versa but conventionally:
- A sparse network is a network where a large number of its parameters where **replaced by zeros**.
- A pruned network is a network where a large number of its parameters were **removed**.

Each method has its own advantages and disadvantages; sparsifying keeps the original shape of the model and the feature maps intact, introducing minimal degradation, it is also very convenient when dealing with complex models. While pruning is will typically require fine-tuning or retraining because of the degradation caused by the ablation (e.g. destruction of the feature maps) but however, pruning will also introduce massive speed up across all types of hardwares (even GPUs).

- The snippet of code you provided does calculate the **sparsity ratio** of a given model.
- Your code returns 9.9e-5 meaning that 0.0099 % of your model's parameters are exactly equal to zero which is normal for a non-sparse model.
- This repo not only sparsify but also **prune** the model.
- To calculate the pruning ratio, you simply have to divide the number of the removed parameters by the number of the total parameters e.g.
![image](https://github.com/VainF/Torch-Pruning/assets/97101162/b8089fcc-8644-4c65-b26d-0bb872b2de19)
((0.036905 - 0.009347) / 0.036905) * 100 = 82.8 % of the parameters have been **removed**.

TL;DR
**Pruned model = sparse model + removal of parameters equal to zero**    (in most workflows).
I hope this was clear,
Don't hesitate to ask any more questions.

**Comment by ajithkumarmcw** [comment]
okay thanks for a clear explanation
